Migrate to MV3 and Modern Redux

median-dxz opened the pull request:
**Note: This PR is currently undergoing beta testing**
 
This will close issues: #1655 #1634
 
So, this is quite a big PR. I've already outlined my changes in [#10](https://github.com/median-dxz/Cookie-AutoDelete-MV3/issues/10), and there are explanations in the commit messages as well. Here, I just want to clarify some of the most important points:
 
---
 
**Q: Why is this PR so large?**  
A: Because I migrated **all the dependencies** to their latest major versions, which introduced a lot of breaking changes, and I had to update the corresponding test suites. These changes cover build tooling, state management, UI styles, TypeScript types, and more.
 
Most of the changes focus on organizing imports, refining TypeScript types, and adapting unit tests. The other major changes include modifications to configuration files and a refactoring of Redux logic. Since the latest version of `@reduxjs/toolkit` is being used, this refactoring is necessary.

---
 
**Q: What does the MV3 migration include?**  
A: In summary, the main changes are:
1. Changed background scripts to Service Workers and modified `browserDetect` so it can run in the SW context
2. Migrated from `browserAction` to `action`
3. Fields change in `manifest.json`
4. Replaced `tabs.executeScript` with `scripting.executeScript`
5. DX: Migrated all TS types to `@types/webextension-polyfill`
6. Fixed: Ensure the timeout is now always triggered in the Service Worker environment
7. Fixed: Fix duplicate menu creation errors
8. Updated: Register event listeners synchronously in the background script

Also, a small reminder: the extension does not actually use `webRequest`, so there is no need to deal with it.
 
 Refer to: [1634#comment](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/1634#issuecomment-2452797529)

---
 
**Q: Has the extension's business logic changed? Is the configuration still compatible?**  
A: Other than not using `tabs.create` to open the options page anymore (which means the config page will now default to the welcome screen), the business logic remains unchanged and the configuration is fully compatible.
 
---
 
**Q: Have the unit tests passed?**  
A: Yes, they have.
 
---
 
**Q: What about user testing?**  
A: I have tested it myself on the latest version of Chrome and found no issues so far. I haven't tested it on Firefox or other extension environments yet.
 
---
 
If you have any questions, feel free to discuss them with me in this PR!

dependabot[bot] committed:
Bump braces, jest, jest-when and ts-jest

---
updated-dependencies:
- dependency-name: braces
  dependency-version: 3.0.3
  dependency-type: indirect
- dependency-name: jest
  dependency-version: 30.0.4
  dependency-type: direct:development
- dependency-name: jest-when
  dependency-version: 3.7.0
  dependency-type: direct:development
- dependency-name: ts-jest
  dependency-version: 29.4.0
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>

median-dxz committed:
Update README.md

median-dxz committed:
chore: change release link

median-dxz committed:
Merge pull request #9 from median-dxz/dependabot/npm_and_yarn/multi-c8c3916126

Bump braces, jest, jest-when and ts-jest

median-dxz committed:
chore: update dependencies and devDependencies to latest versions

median-dxz committed:
chore: migrate ESLint configuration to flat config

median-dxz committed:
refactor: convert IconButton to adapt React 19

median-dxz committed:
chore: migrate jest configuration to adapt jest v30

median-dxz committed:
fix: correct logical condition in filterSiteData function

median-dxz committed:
refactor: migrate to react 19

median-dxz committed:
fix: specify generic types for store parameter in StoreUser.init method

median-dxz committed:
refacor: migrate to mv3, fix typing issues (not include tests)

- utilize `webextension-polyfill`
- refactor global declarations

median-dxz committed:
refactor: migrate background redux logic to Redux Toolkit

- take full advantage of redux toolkit
- use redux-logger middleware
- replace redux-webext with webext-redux
- use createAction instead of defining action manually
- better thunk action typo
- refactor actions file organization
- use action creators instead of make actions manaually

median-dxz committed:
refactor: migrate redux logic in React to hooks API

- Add the hooks eslint plugin
- For React FC, use hooks to access dispatch and state

median-dxz committed:
fix: correct BrowserActionService typo

median-dxz committed:
fix: not pass proxy store correctly

median-dxz committed:
perf: add shadowEqual to some selectors

median-dxz committed:
fix: incorrect payload received by background

median-dxz committed:
fix: correct the typos in test

- Properly mock `browser`
- Use generic parameters to provide more precise typings for  `JestSpyObject` and `generateSpies`
- Adapt new browser api
- Adapt to new Jest api

median-dxz committed:
refactor: use action creators instead of manually constructing actions in tests

median-dxz committed:
fix：some human errors introduced during code migration

median-dxz committed:
refactor: simplify alias definition by a wrapper function

median-dxz committed:
fix: some matchers would be overridden by accident using `jest-when`

median-dxz committed:
fix: correct logical condition in filterSiteData function

-  Revert changes from a36524dff3 and fix it again

median-dxz committed:
fix: use namespace import to make spy work

median-dxz committed:
perf & fix: organize configurations, update incorrect and outdated items, enable ESM support by default

median-dxz committed:
toolchain: migrate build toolchain to Vite and CRXJS in development environment

- Treat `browserDetect` as a utility module

median-dxz committed:
style & fix: fix pupup display issues

median-dxz committed:
style & fix: fix the ActivityTable styles

median-dxz committed:
update: MV3 migration: use `scripting.executeScript`

See more: https://developer.chrome.com/blog/crx-scripting-api

median-dxz committed:
perf: avoid accessing root state in React components directly

- Transform some functions to use `createAsyncThunk`

median-dxz committed:
chore: suppress some eslint errors

- remove unused import
- remove unnessary state copy in tests

median-dxz committed:
refactor: replace `state` with `isFirefox` in returnOptionalCookieAPIAttributes calls

median-dxz committed:
refactor: ActivityTable: extract subcomponents as standalone components

median-dxz committed:
fix: improve cookie restoration error handling using Promise.allSettled

median-dxz committed:
style & fix: fix Options Page style

median-dxz committed:
toolchain: integrate new development workflow with legacy build process

median-dxz committed:
feat: support different browser environments in development mode

median-dxz committed:
fix: add contextMenus setting check in ContextMenuEvents methods

median-dxz committed:
toolchain: set root dir

median-dxz committed:
fix: validate tab existence before processing actions in onTabUpdate

median-dxz committed:
fix: mistakenly changed output dir to `extension`

median-dxz committed:
style: adjust some gaps

median-dxz committed:
Bump ver to v3.8.3

Cuasars commented:
Hello, Chrome won't let me use the extension, so I've started looking for alternatives to Chrome, Vivaldi. I more or less understand it now, but I'm having problems with that browser's translations. There are pages that it doesn't translate, etc. Brave and some others, there isn't an equally easy and intuitive alternative to cookie autodelete. I hope to have it back soon. Thanks.

median-dxz committed:
fix: suppress ESLint `no-extraneous-class` error

Cuasars commented:
THANK YOU for the work!  👏👏👏👏

median-dxz committed:
feat: add `waitUntil` for activating SW in long tasks

- add setupFakeTimers utility for better timer handling in tests

median-dxz committed:
feat: use `waitUntil` to prevent SW from being deactivated during `setTimeout`

median-dxz committed:
perf & type: remove `any` from dispatch

median-dxz committed:
feat: use `Alarm` to create timer for cleanup delay longer than 60s

median-dxz committed:
fix: move `menuInit` to the `onInstalled` callback

- Remove the `isInitialized` property from `ContextMenuEvents` because it will be reset in the SW environment.

median-dxz committed:
version: v3.8.3-beta.2 release notes

median-dxz committed:
test: should no longer rely on `isInitialized` state

median-dxz committed:
test: spy on `ContextMenus` in SettingService test suites properly

dependabot[bot] committed:
chore(deps-dev): bump vite from 7.0.5 to 7.1.5 (#18)

Bumps [vite](https://github.com/vitejs/vite/tree/HEAD/packages/vite) from 7.0.5 to 7.1.5.
- [Release notes](https://github.com/vitejs/vite/releases)
- [Changelog](https://github.com/vitejs/vite/blob/main/packages/vite/CHANGELOG.md)
- [Commits](https://github.com/vitejs/vite/commits/v7.1.5/packages/vite)

---
updated-dependencies:
- dependency-name: vite
  dependency-version: 7.1.5
  dependency-type: direct:development
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

median-dxz committed:
chore: bump deps

median-dxz committed:
style: adjusted the layout of the `ActivityItem` and fixed some minor styling issues.

- Adjusted icon width after migrating to Fontawesome v7
- Changed `sr-only` to `visually-hidden`

median-dxz committed:
test: correctly mock the APIs under the `action` namespace

median-dxz committed:
migrate: register event listeners synchronously in the background script

median-dxz committed:
style: fix incorrect styles of `SelectInput`

median-dxz committed:
perf: remove redundant return type declaration

median-dxz committed:
version: v3.8.2-beta.2 re-publish

median-dxz committed:
fix: should not use the same Manifest in all builds

median-dxz committed:
fix: wrap `onCookieChanged` with `StoreUser.withStoreReady`

median-dxz committed:
fix: reset session cookie counter correctly

median-dxz committed:
chore bump deps

median-dxz committed:
chore: update eslint config

median-dxz committed:
Revert README for original repo

This reverts commit d0e326fdad845646c0bf21b309ea49f0273753d8 and 97744ff113bc14a683bee9d45afeb2e65228df5e.

median-dxz committed:
fix: update Firefox build process

see: https://github.com/mozilla/web-ext/issues/2532

kennethtran93 commented:
That's a lot more work done than what I had going on on my local machine version (which is still incomplete as I have poor time management).  If all is good I don't mind using this as v4.0 since anything we need to fix at this point is a major version upgrade.  Would most likely be a completely new branch.

Happy Holidays.
